### PR TITLE
Allow renaming child nodes in `_ready`

### PR DESCRIPTION
**Pull request by RedworkDE**
- Implements https://github.com/godotengine/godot/pull/78654#issuecomment-1605755006
- Re-enables previously valid code
- Probably best as 4.2 + cherry-pick

### CODE REVIEW

**Comment by akien-mga** [comment]
@reduz seemed to like that idea when proposed by @kleonc in https://github.com/godotengine/godot/issues/77610#issuecomment-1606274916. CC @Sauermann 

Agree with keeping this for 4.2 and possibly a cherrypick for 4.1.1 to lift the restriction, so we don't take risks now.

File: core/templates/hash_map.h [hunk 1]
```diff
@@ -353,6 +353,38 @@ class HashMap {
 		return true;
 	}
 
+	bool replace_key(const TKey &p_old_key, const TKey &p_new_key) {
+		uint32_t pos = 0;
+		bool exists = _lookup_pos(p_old_key, pos);
```
Comment (Sauermann) [inline core/templates/hash_map.h:358]: Should it be checked, if the new key exists?
```c++
if (_lookup_pos(p_new_key, pos)) { return; }
```

Comment (RedworkDE) [reply core/templates/hash_map.h]: Was happening 3 lines down, but I moved the elements access below it.

Comment (Sauermann) [reply core/templates/hash_map.h]: Please note, that I talk about `p_new_key` and not `p_old_key`.

Comment (RedworkDE) [reply core/templates/hash_map.h]: Reading is hard. I added that check and also an early return for if the old and new key are equal.

**Review by reduz — APPROVED** [review]

**Comment by reduz** [comment]
Looks good, great approach!

File: scene/main/node.cpp [hunk 2]
```diff
@@ -1149,9 +1148,8 @@ void Node::set_name(const String &p_name) {
 	data.name = name;
 
 	if (data.parent) {
-		data.parent->data.children.erase(old_name);
 		data.parent->_validate_child_name(this, true);
-		data.parent->data.children.insert(data.name, this);
+		ERR_FAIL_COND(!data.parent->data.children.replace_key(old_name, data.name));
```
Comment (kleonc) [inline scene/main/node.cpp:1152]: I think the preferred code style here is to make a mutating call like that be explicitely separated:
```suggestion
		bool success = data.parent->data.children.replace_key(old_name, data.name);
		ERR_FAIL_COND(!success);
```
Example from the same file:
https://github.com/godotengine/godot/blob/afc5fa14adce7676746662751e63eab2d2ad4420/scene/main/node.cpp#L1439-L1440

File: core/templates/hash_map.h [hunk 3]
```diff
@@ -353,6 +353,38 @@ class HashMap {
...
+			next_pos = fastmod((pos + 1), capacity_inv, capacity);
+		}
+		hashes[pos] = EMPTY_HASH;
+		elements[pos] = nullptr;
+		// _insert_with_hash will increment this again.
+		num_elements--;
+
+		// Update the HashMapElement with the new key and reinsert it.
+		const_cast<TKey &>(element->data.key) = p_new_key;
+		uint32_t hash = _hash(p_new_key);
```
Comment (kleonc) [inline core/templates/hash_map.h:382]: This is already calculated in `_lookup_pos(p_new_key, pos)` so there's a room for optimizing it.

**Review by kleonc — APPROVED** [review]
Overall the code looks good to me. :+1: AFAICT should be good as is, but left some notes.

File: core/templates/hash_map.h [hunk 4]
```diff
@@ -353,6 +353,38 @@ class HashMap {
...
+		ERR_FAIL_COND_V(!_lookup_pos(p_old_key, pos), false);
+		HashMapElement<TKey, TValue> *element = elements[pos];
+
+		// Delete the old entries in hashes and elements.
+		const uint32_t capacity = hash_table_size_primes[capacity_index];
+		const uint64_t capacity_inv = hash_table_size_primes_inv[capacity_index];
+		uint32_t next_pos = fastmod((pos + 1), capacity_inv, capacity);
+		while (hashes[next_pos] != EMPTY_HASH && _get_probe_length(next_pos, hashes[next_pos], capacity, capacity_inv) != 0) {
+			SWAP(hashes[next_pos], hashes[pos]);
+			SWAP(elements[next_pos], elements[pos]);
```
[the remark this answers is not in the document]
Comment (RedworkDE) [reply core/templates/hash_map.h:371]: This is copied as-is from `erase` and would rather keep it simple for this PR, same for the duplicate hash computation below (also some operations can compute the hash / lookup the element 3 or 4 times, so hash map really deserves a dedicated optimization PR with measurements etc)

[the remark this answers is not in the document]
Comment (kleonc) [reply core/templates/hash_map.h:371]: Agree. As mentioned, these are just some notes/thoughts. :upside_down_face:

**Comment by YuriSizov** [comment]
Thanks!

**Pull request merged**